traffic_ops/testing/api/v3/deliveryserviceservers_test.go

```diff
@@ -17,21 +17,339 @@ package v3
+		t.Fatalf("Failed to create new Profile in CDN #%d: %v", cdnID, err)
+	}
+	profiles, _, err := TOSession.GetProfileByNameWithHdr(profile.Name, nil)
+	if err != nil {
+		t.Fatalf("Could not fetch Profile '%s' after creation: %v", profile.Name, err)
+	}
+	if len(profiles) != 1 {
+		t.Fatalf("Expected exactly one Profile named '%s'; got: %d", profile.Name, len(profiles))
+	}
+	profile = profiles[0]
+
+	physLocs, _, err := TOSession.GetPhysLocationsWithHdr(nil, nil)
+	if err != nil {
+		t.Fatalf("Could not fetch Physical Locations: %v", err)
+	}
+	if len(physLocs) < 1 {
+		t.Fatal("Need at least one Physical Location")
+	}
+	physLoc := physLocs[0]
+
+	statuses, _, err := TOSession.GetStatusesWithHdr(nil)
+	if err != nil {
+		t.Fatalf("Could not fetch Statuses: %v", err)
+	}
+	if len(statuses) < 1 {
+		t.Fatal("Need at least one Status")
+	}
+
+	var statusID int
+	var badStatusID int
+	found := false
+	foundBad := false
+	for _, status := range statuses {
+		if status.Name == "ONLINE" || status.Name == "REPORTED" {
+			statusID = status.ID
+			found = true
+			if foundBad {
+				break
+			}
+		} else {
+			badStatusID = status.ID
+			foundBad = true
+			if found {
+				break
+			}
+		}
+	}
+	if !found || !foundBad {
+		t.Fatal("Need at least one status with the name 'ONLINE' or 'REPORTED' and at least one status with neither of those names")
+	}
+
+	types, _, err = TOSession.GetTypesWithHdr(nil, "server")
+	if err != nil {
+		t.Fatalf("Could not fetch Types: %v", err)
+	}
+	if len(types) < 1 {
+		t.Fatal("Need at least one Type with 'useInTable' of 'server' to create a server")
+	}
+	serverTypeID := types[0].ID
+
+	server := tc.ServerV30{
+		CommonServerProperties: tc.CommonServerProperties{
+			CachegroupID:   &cacheGroupID,
+			CDNID:          &cdnID,
+			DomainName:     &xmlid,
+			HostName:       &xmlid,
+			HTTPSPort:      util.IntPtr(443),
+			PhysLocationID: &physLoc.ID,
+			ProfileID:      &profile.ID,
+			RevalPending:   new(bool),
+			StatusID:       &statusID,
+			TCPPort:        util.IntPtr(80),
+			TypeID:         &serverTypeID,
+			UpdPending:     new(bool),
+		},
+		Interfaces: []tc.ServerInterfaceInfo{
+			{
+				Name: "eth0",
+				IPAddresses: []tc.ServerIPAddress{
+					{
+						Address:        "198.51.100.1",
+						ServiceAddress: true,
+					},
+				},
+			},
+		},
+	}
+	alerts, _, err = TOSession.CreateServerWithHdr(server, nil)
+	if err != nil {
+		t.Fatalf("Failed to create server: %v - alerts: %s", err, strings.Join(alerts.ToStrings(), ", "))
+	}
+	params := url.Values{}
+	params.Set("hostName", *server.HostName)
+	servers, _, err := TOSession.GetServersWithHdr(&params, nil)
+	if err != nil {
+		t.Fatalf("Could not fetch server after creation: %v", err)
+	}
+	if len(servers.Response) != 1 {
+		t.Fatalf("Expected exactly 1 server with hostname '%s'; got: %d", *server.HostName, len(servers.Response))
+	}
+	server = servers.Response[0]
+	if server.ID == nil {
+		t.Fatal("Server had null/undefined ID after creation")
+	}
```

Comment:
Because then I have to depend on the structure of `tc-fixtures` and changes therein could break the test. Like how `AssignTestDeliveryService` takes the first server and first DS in the fixture data and assigns them to one another, then if I put mine in as the first ds and server then I have to go change _that_ test to use some other hard-coded data. And If I put my ds and server last in the fixture data, then any time someone adds or remove a DS or server from the fixture data this test breaks.

This is very annoying, but it just seemed way less brittle than hard-coding a position in the fixture data.

Reply:
You don't have to depend on a particular index in tc-fixtures.json. I agree that is bad practice and should probably never be done. We should generally always be free to _add_ data to tc-fixtures.json, and any test that prevents that must be fixed. However, I think it is perfectly acceptable to add new data to tc-fixtures.json and reference primary keys (e.g. xml_id) in the test, because generally people should not _change_ existing objects in tc-fixtures.json (at least not significantly). It is usually safer to try to create the test from existing data (if possible, without changing it), or just add new data to tc-fixtures.json.

Reply:
I did that, also made a note that the tests expect all newly added Delivery Services in the fixture data to belong to the `tenant1` Tenant, or the tenancy tests will break